presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownTableConstraints.java

```diff
@@ -0,0 +1,154 @@
+
+public class PushDownTableConstraints
+        implements Rule
+{
+    private final Metadata metadata;
+
+    public PushDownTableConstraints(Metadata metadata)
+    {
+        this.metadata = metadata;
+    }
+
+    @Override
+    public Optional<PlanNode> apply(PlanNode node, Lookup lookup, PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator, Session session)
+    {
+        if (!(node instanceof FilterNode) || !(lookup.resolve(((FilterNode) node).getSource()) instanceof TableScanNode)) {
+            return Optional.empty();
+        }
+
+        FilterNode filter = (FilterNode) node;
+        Expression predicate = filter.getPredicate();
+        // don't include non-deterministic predicates
+        Expression deterministicPredicate = stripNonDeterministicConjuncts(predicate);
+
+        DomainTranslator.ExtractionResult decomposedPredicate = DomainTranslator.fromPredicate(
+                metadata,
+                session,
+                deterministicPredicate,
+                symbolAllocator.getTypes());
+
+        TableScanNode tableScan = (TableScanNode) lookup.resolve(filter.getSource());
+        TupleDomain<ColumnHandle> simplifiedConstraint = decomposedPredicate.getTupleDomain()
+                .transform(tableScan.getAssignments()::get)
+                .intersect(tableScan.getCurrentConstraint());
+
+        Map<ColumnHandle, Symbol> assignments = ImmutableBiMap.copyOf(tableScan.getAssignments()).inverse();
+
+        // Layouts will be returned in order of the connector's preference
+        List<TableLayoutResult> layouts = metadata.getLayouts(
+                session, tableScan.getTable(),
+                new Constraint<>(simplifiedConstraint, bindings -> true),
+                Optional.of(tableScan.getOutputSymbols().stream()
+                        .map(tableScan.getAssignments()::get)
+                        .collect(toImmutableSet())));
+
+        if (layouts.isEmpty()) {
+            return Optional.of(new ValuesNode(idAllocator.getNextId(), tableScan.getOutputSymbols(), ImmutableList.of()));
+        }
+
+        // Filter out layouts that cannot supply all the required columns
+        layouts = layouts.stream()
+                .filter(layoutHasAllNeededOutputs(tableScan))
+                .collect(toList());
+        checkState(!layouts.isEmpty(), "No usable layouts for %s", tableScan);
+
+        // At this point we have no way to choose between possible layouts, just take the first one
+        TableLayoutResult layout = layouts.get(0);
+
+        PlanNode rewrittenPlan = new TableScanNode(
+                tableScan.getId(),
+                tableScan.getTable(),
+                tableScan.getOutputSymbols(),
+                tableScan.getAssignments(),
+                Optional.of(layout.getLayout().getHandle()),
+                simplifiedConstraint.intersect(layout.getLayout().getPredicate()),
+                Optional.ofNullable(tableScan.getOriginalConstraint()).orElse(predicate));
```

Comment:
`tableScan.getOriginalConstraint()` bears this comment:

> "This field exists for the sole purpose of being able to print the original predicates"

Are you sure you want to use it?